### PR TITLE
feat: add duplicate_name_check expectation

### PR DESCRIPTION
**Pull request by tombrooks248**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## What

Adds a `duplicate_name_check` dataset expectation, which flags names reused by several entities within the same provision — the same dataset and organisation.

## Why

Part 3 of digital-land#598. The same `name` text is reused across distinct entities, so the name stops identifying anything — usually one feature split into several polygons whose parts were never distinguished. For example one authority used `"Conservation_Area.TAB"` for 27 conservation areas, and another used `"District Centre related A4D"` for 364 article-4-direction areas.

Grouping is per organisation rather than dataset-wide because the published dataset is every LPA's file concatenated together; `"City Centre"` appears once each in seven different councils, and none of them can act on that. Matching reuses the existing `_normalise_name` helper, so case and spacing variants group together while the raw spelling is what gets reported.

## Related Tickets & Documents

- [Ticket Link](https://github.com/digital-land/config/issues/2911)
- Related Issue #
- Closes #

## Testing

5 integration tests covering per-organisation scoping, the threshold parameter, normalisation, blank exclusion and the failure structure. 105 expectations tests pass, black and flake8 clean.

Checked against currently published data: 466 duplicate-name groups for `article-4-direction-area` and 178 for `conservation-area`.

## This change is inert
This change is inert on its own. The operation only runs for a dataset once a row naming it is added to that collection's `expect.csv`, so merging this alters no pipeline output. Opting in `article-4-direction-area` and `conservation-area` follows in a separate config PR.

## Added/updated tests?
_We encourage you to keep the code coverage percentage at 80% and above. Please refer to the [Digital Land Testing Guidance](https://digital-land.github.io/technical-documentation/development/testing-guidance/) for more information._

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## Summary by CodeRabbit

- **New Features**
  - Added dataset validation to detect repeated entity names.
  - Matching is case- and whitespace-insensitive, while blank names are excluded.
  - Supports configurable thresholds for identifying duplicates.
  - Results are consistently ordered and include relevant entity references.

- **Tests**
  - Added coverage for duplicate detection, dataset and organisation scoping, name normalisation, blank-name handling, and configurable thresholds.

### CODE REVIEW

File: digital_land/expectations/operations/dataset.py [hunk 1]
```diff
@@ -641,6 +641,83 @@ def name_is_a_placeholder_check(conn, listed_building_path, placeholders=None):
...
+            and trim(name) != ''
+    """
+    rows = conn.execute(query).fetchall()
+
+    groups = {}
+    for entity, reference, name, organisation_entity in rows:
+        key = (organisation_entity, _normalise_name(name))
+        # keep the first raw spelling seen: the normalised key is not worth showing to
+        # an LPA, and two rows differing only in case would otherwise look identical
+        group = groups.setdefault(key, {"name": name, "entities": [], "references": []})
```
[the remark this answers is not in the document]
Comment (tombrooks248) [reply digital_land/expectations/operations/dataset.py:690]: Good idea - done

File: digital_land/expectations/operations/dataset.py [hunk 2]
```diff
@@ -641,6 +641,90 @@ def name_is_a_placeholder_check(conn, listed_building_path, placeholders=None):
     return result, message, details
 
 
+def duplicate_name_check(conn, threshold=2):
+    """
```
Comment (gibahjoe) [inline digital_land/expectations/operations/dataset.py:645]: The function doc says it checks within a provision however, the connection could contain data from different datasets (its a public function) and the grouping process does not  group by dataset. 

This means you could run this check  across different provisions and end up with duplicates.

File: tests/integration/expectations/operations/test_dataset.py [hunk 3]
```diff
@@ -968,3 +969,118 @@ def test_name_is_a_placeholder_check_single_string_placeholder(
...
+        # same name twice in one org, should be flagged
+        (1, "A4D-1", "Site allocation NSP30", "600001"),
+        (2, "A4D-2", "Site allocation NSP30", "600001"),
+        # three in one org, should be flagged with a count of 3
+        (3, "A4D-3", "District Centre related A4D", "600001"),
+        (4, "A4D-4", "District Centre related A4D", "600001"),
+        (5, "A4D-5", "District Centre related A4D", "600001"),
+        # used once, should not be flagged
+        (6, "A4D-6", "Queen's Park", "600001"),
+    ]
```
Comment (gibahjoe) [inline tests/integration/expectations/operations/test_dataset.py:993]: Add a test case where duplicate names exist across different datasets, to ensure the behaviour remains consistent.

**Review by gibahjoe — DISMISSED** [review]
More of a nitpick. 

Also consider running the grouping in the connection since you already have it which will probably be beneficial for humongous rows.

File: digital_land/expectations/operations/dataset.py (hunk 2 above)
Comment (tombrooks248) [reply digital_land/expectations/operations/dataset.py:645]: Good idea - yea I have fixed the code to make it do what it says it does, and hopefully make it more useful if someone else wants to re-use the code in future.

File: tests/integration/expectations/operations/test_dataset.py (hunk 3 above)
Comment (tombrooks248) [reply tests/integration/expectations/operations/test_dataset.py:993]: Yes and added a relevant test for it.

**Comment by tombrooks248** [comment]
> More of a nitpick.
> 
> Also consider running the grouping in the connection since you already have it which will probably be beneficial for humongous rows.

On this - I looked into it and because its running the python function `_normalise_name(name)` which does some things in python (regex, internal trimming, handling tabs and newlines) which would be be very complicated/ possible not doable in SQL, I don't think its worth refactoring into SQL. But I agree in general doing this stuff in SQL is better when that is possible.

**Review by gibahjoe — APPROVED** [review]
LGTM
